### ISSUE TITLE
Fails to compile with nightly 2018-03-29

### ISSUE THREAD

**Issue by phansch** [vorner/tokio-jsonrpc#7]
Hi,

I have been trying to reproduce https://github.com/rust-lang-nursery/rust-clippy/issues/1515 but it looks like tokio-jsonrpc doesn't compile with the current nightly? I'm not sure since when that's the case, though.

Here is the error output:

```
Compiling tokio-jsonrpc v0.9.1 (file:///home/phansch/code/repros/tokio-jsonrpc)
error[E0277]: the trait bound `std::result::Result<_, _>: futures::Future` is not satisfied
   --> src/endpoint.rs:643:23
    |
643 |         let cleaner = unfold((), move |_| -> Option<Result<_, _>> {
    |                       ^^^^^^ the trait `futures::Future` is not implemented for `std::result::Result<_, _>`
    | 
    = note: required because of the requirements on the impl of `futures::Future` for `std::option::Option<std::result::Result<_, _>>`
    = note: required by `futures::stream::unfold`

error[E0277]: the trait bound `std::result::Result<(std::option::Option<_>, ()), _>: futures::Future` is not satisfied
   --> src/endpoint.rs:643:23
    |
643 |           let cleaner = unfold((), move |_| -> Option<Result<_, _>> {
    |  _______________________^                 
644 | |             let mut idmap = idmap_cloned.borrow_mut();
645 | |             debug!(logger_cloned, "Dropping unanswered RPCs (EOS)"; "outstanding" => idmap.len());
646 | |             idmap.clear();              
...   |
650 | |             Some(Ok((None, ())))        
651 | |         });                             
    | |__________^ the trait `futures::Future` is not implemented for `std::result::Result<(std::option::Option<_>, ()), _>`
    |                                           
    = note: required because of the requirements on the impl of `futures::Future` for `std::option::Option<std::result::Result<(std::option::Option<_>, ()), _>>`
    = note: required because of the requirements on the impl of `futures::IntoFuture` for `std::option::Option<std::result::Result<(std::option::Option<_>, ()), _>>`
    = note: required by `futures::stream::Unfold`
                                                
error[E0277]: the trait bound `std::result::Result<(std::option::Option<_>, ()), _>: futures::Future` is not satisfied
   --> src/endpoint.rs:656:14
    |                                           
656 |             .chain(cleaner)               
    |              ^^^^^ the trait `futures::Future` is not implemented for `std::result::Result<(std::option::Option<_>, ()), _>`
    |
    = note: required because of the requirements on the impl of `futures::Future` for `std::option::Option<std::result::Result<(std::option::Option<_>, ()), _>>`

error[E0599]: no method named `select` found for type `futures::stream::Chain<futures::stream::Map<futures::stream::SplitStream<Connection>, fn(std::result::Result<message::Message, message::B
roken>) -> std::option::Option<std::result::Result<message::Message, message::Broken>> {std::option::Option<std::result::Result<message::Message, message::Broken>>::Some}>, futures::stream::Un
fold<(), [closure@src/endpoint.rs:643:34: 651:10 idmap_cloned:_, logger_cloned:_, ctl_clone:_], std::option::Option<std::result::Result<(std::option::Option<_>, ()), _>>>>` in the current scop
e
   --> src/endpoint.rs:657:14
    |
657 |             .select(terminator)
    |              ^^^^^^
    |
    = note: the method `select` exists but the following trait bounds were not satisfied:
            `futures::stream::Chain<futures::stream::Map<futures::stream::SplitStream<Connection>, fn(std::result::Result<message::Message, message::Broken>) -> std::option::Option<std::result
::Result<message::Message, message::Broken>> {std::option::Option<std::result::Result<message::Message, message::Broken>>::Some}>, futures::stream::Unfold<(), [closure@src/endpoint.rs:643:34:
651:10 idmap_cloned:_, logger_cloned:_, ctl_clone:_], std::option::Option<std::result::Result<(std::option::Option<_>, ()), _>>>> : futures::Stream`
            `&mut futures::stream::Chain<futures::stream::Map<futures::stream::SplitStream<Connection>, fn(std::result::Result<message::Message, message::Broken>) -> std::option::Option<std::r
esult::Result<message::Message, message::Broken>> {std::option::Option<std::result::Result<message::Message, message::Broken>>::Some}>, futures::stream::Unfold<(), [closure@src/endpoint.rs:643
:34: 651:10 idmap_cloned:_, logger_cloned:_, ctl_clone:_], std::option::Option<std::result::Result<(std::option::Option<_>, ()), _>>>> : futures::Future`
            `&mut futures::stream::Chain<futures::stream::Map<futures::stream::SplitStream<Connection>, fn(std::result::Result<message::Message, message::Broken>) -> std::option::Option<std::r
esult::Result<message::Message, message::Broken>> {std::option::Option<std::result::Result<message::Message, message::Broken>>::Some}>, futures::stream::Unfold<(), [closure@src/endpoint.rs:643
:34: 651:10 idmap_cloned:_, logger_cloned:_, ctl_clone:_], std::option::Option<std::result::Result<(std::option::Option<_>, ()), _>>>> : futures::Stream`

error: aborting due to 4 previous errors
            `futures::stream::Chain<futures::stream::Map<futures::stream::SplitStream<Connection>, fn(std::result::Result<message::Message, message::Broken>) -> std::option::Option<std::result
::Result<message::Message, message::Broken>> {std::option::Option<std::result::Result<message::Message, message::Broken>>::Some}>, futures::stream::Unfold<(), [closure@src/endpoint.rs:643:34:
651:10 idmap_cloned:_, logger_cloned:_, ctl_clone:_], std::option::Option<std::result::Result<(std::option::Option<_>, ()), _>>>> : futures::Stream`
            `&mut futures::stream::Chain<futures::stream::Map<futures::stream::SplitStream<Connection>, fn(std::result::Result<message::Message, message::Broken>) -> std::option::Option<std::r
esult::Result<message::Message, message::Broken>> {std::option::Option<std::result::Result<message::Message, message::Broken>>::Some}>, futures::stream::Unfold<(), [closure@src/endpoint.rs:643
:34: 651:10 idmap_cloned:_, logger_cloned:_, ctl_clone:_], std::option::Option<std::result::Result<(std::option::Option<_>, ()), _>>>> : futures::Future`
            `&mut futures::stream::Chain<futures::stream::Map<futures::stream::SplitStream<Connection>, fn(std::result::Result<message::Message, message::Broken>) -> std::option::Option<std::r
esult::Result<message::Message, message::Broken>> {std::option::Option<std::result::Result<message::Message, message::Broken>>::Some}>, futures::stream::Unfold<(), [closure@src/endpoint.rs:643
:34: 651:10 idmap_cloned:_, logger_cloned:_, ctl_clone:_], std::option::Option<std::result::Result<(std::option::Option<_>, ()), _>>>> : futures::Stream`

error: aborting due to 4 previous errors
```

**Comment by vorner** [comment]
Hello

Looking at it, this is caused by a breaking change in futures library: https://github.com/rust-lang-nursery/futures-rs/issues/904.

I'll wait a while if it resolves itself before I start thinking about how to fix it on my end (I'll have to decide about dropping support for older versions, not supporting the new one or using another modifier).

For now, can you just put `futures = "=0.1.18"` into `Cargo.toml` for the experiments?

**Comment by phansch** [comment]
Thanks! Using futures 0.1.18 made it compile and clippy is not crashing anymore inside the `src` directory :+1:

**Comment by vorner** [comment]
Fixed in the futures crate in 0.1.21.
